crates/core/src/topology.rs

```diff
@@ -134,43 +143,376 @@ impl TopologyManager {
 
         let best_location = match density_map.get_max_density() {
             Ok(location) => {
-                tracing::debug!("Max density found at location: {:?}", location);
+                debug!("Max density found at location: {:?}", location);
                 location
             }
             Err(_) => {
-                tracing::debug!(
+                debug!(
                     "An error occurred while getting max density, falling back to random location"
                 );
-                self.random_location()
+                this_peer_location.clone()
             }
         };
         Ok(best_location)
     }
 
-    /// Generates a random location that is close to the current peer location with a small
-    /// world distribution.
-    fn random_location(&self) -> Location {
-        tracing::debug!("Generating random location");
-        let distance = random_link_distance(Distance::new(RANDOM_CLOSEST_DISTANCE));
-        let location_f64 = if rand::random() {
-            self.this_peer_location.as_f64() - distance.as_f64()
+    pub fn update_limits(&mut self, limits: Limits) {
+        self.limits = limits;
+    }
+
+    pub(crate) fn report_resource_usage(
+        &mut self,
+        attribution: &AttributionSource,
+        resource: ResourceType,
+        amount: f64,
+        at_time: Instant,
+    ) {
+        if let Some(creation_time) = self.source_creation_times.get(attribution) {
+            if at_time < *creation_time {
+                self.source_creation_times
+                    .insert(attribution.clone(), at_time);
+            }
         } else {
-            self.this_peer_location.as_f64() + distance.as_f64()
+            self.source_creation_times
+                .insert(attribution.clone(), at_time);
+        }
+
+        self.meter.report(attribution, resource, amount, at_time);
+    }
+
+    /// Record an outbound request to a peer, along with the target Location of that request
+    pub(crate) fn report_outbound_request(&mut self, peer: PeerKeyLocation, target: Location) {
+        self.request_density_tracker.sample(target);
+        self.outbound_request_counter.record_request(peer);
+    }
+
+    /// Calculate total usage for a resource type extrapolating usage for resources that
+    /// are younger than [SOURCE_RAMP_UP_DURATION]
+    fn extrapolated_usage(&mut self, resource_type: &ResourceType, now: Instant) -> Usage {
+        let function_span = span!(Level::DEBUG, "extrapolated_usage_function");
+        let _enter = function_span.enter();
+
+        debug!("Function called");
+        debug!("Resource type: {:?}", resource_type);
+        debug!("Current time: {:?}", now);
+
+        let mut total_usage: Rate = Rate::new_per_second(0.0);
+        let mut usage_per_source: HashMap<AttributionSource, Rate> = HashMap::new();
+
+        // Step 1: Collect data
+        let collect_data_span = span!(Level::DEBUG, "collect_data");
```

Comment:
an other way of doing it is rapping each of the steps with an async block and call `instrument(span!(...))` into it instead of explicitly havign to enter and drop the span


eg:
```
async  {
 let usage_data = vec![];
 ...
 }.instrument(debug_span!("collect_data")).await;

Reply:
Hmm, not sure about this because `extrapolate_usage()` isn't otherwise an async function so .await doesn't work

Reply:
Ah yes, nvm